google-apis-common/src/lib.rs

```diff
@@ -777,86 +769,14 @@ pub async fn get_body_as_string(res_body: &mut hyper::Body) -> String {
     res_body_string.to_string()
 }
 
-// TODO: Simplify this to Option<String>
-type TokenResult = std::result::Result<Option<String>, oauth2::Error>;
```

Comment:
Something I'd like to see in the docs is how to obtain the reason for authentication failure. Especially when using `yup-oauth2`, it's easy to botch auth and knowing what failed exactly helps to resolve such issues.

My suggestion is to either keep this as `Result<Option<` but turn the error into `Box<dyn std::error::Error + Send + Sync + 'static>`, or to adjust the `GetToken` trait to query errors for the past `get_token()` call, which would then be used by the generated code to report errors.

To me it seems that leaving this as `Result<Option` is much more straightforward to not degenerate information. From my experience that's the way to go and the best way to indicate that `something can fail, and it might not exist, and checking for existence might fail itself` (hence, `Result<Option<` instead of `Option<Result<`).

Reply:
This is a bit awkward, since I think that the return type should either be `Result<String, Err>` or `Option<String>`. I think `Result<String, Err>` makes sense to make debugging easier.

There are only really two use-cases for `Result<Option<_>, _>`:
1. You want to use your own `Delegate::token` implementation without describing an error, so you return `Ok(None)` (not really important)
2. You want to use `NoToken`. However, calling `get_token()` on `NoToken` likely indicates an incorrect implementation - either the library is calling `get_token` in a no-auth context, or the user is using `NoToken` in an auth context

On further thought, I feel like `NoToken::get_token()` should either panic (not entirely unreasonable, since calling it is an implementation bug) or return an error.

Reply:
After thinking about `Result<Option<_>, _>` a bit further, the main use-case for something like this is in APIs which may or may not require authentication based on the resource being accessed. 

For example, you might want to fetch metadata associated with publicly available Google Drive files. In this case, `Ok(None)` should indicate that the API request should proceed without a token (and any errors that result should be bubbled up). Supporting this would require some adjustments to how tokens are used, but should easy to do.

The other case that would be useful (barely) is if a user wants to call `Delegate::token` without defining their own error type. However, this doesn't seem like a particularly important case, and interferes with the other case described above, which is entirely possible.

Otherwise, if there / are no APIs which may or may not require an API key, then it would make sense for `NoToken::get_token` to simply return an error (or panic, as calling it would be an implementation bug).

Reply:
Example usecase for `Result<Option<_>, _>`: https://issuetracker.google.com/issues/168687448?pli=1

Reply:
Concrete usecase for `Result<Option<_>, _>` - blogger v3: https://developers.google.com/blogger/docs/3.0/reference/blogs/get

A request does not require authentication for public resources